Exercise 1: Operations Excellence
1. Write the optimization models for each plant, i.e., make sure to detail the decision variables, the objective function and the constraints.
2. Provide the optimal solutions and profit levels for each factory. Give the values of the decision variables and the optimal objective function value.

In [17]:
import cvxpy as cp

# BIC Site: x1 (standard), x2 (deluxe)
x1 = cp.Variable(integer=True)
x2 = cp.Variable(integer=True)


# Define the objective function
objective = cp.Maximize(10*x1  + 15*x2)

# Define the constraints
constraints = [
    4*x1 + 2*x2 <= 80,
    2*x1 + 5*x2 <= 60,
    4*(x1 + x2) <= 120,
    x1 >= 0,
    x2 >= 0]

# Define the optimization problem
problem = cp.Problem(objective, constraints)

# Solve the problem
problem.solve()

# Print the status of the solution
print("Status:", problem.status)

# Print the optimal values of the variables
print("x1 = {:.2f}".format(x1.value))
print("x2 = {:.2f}".format(x2.value))


# Define the profit per unit for each product
profit_per_unit = {
    'x1': 10,
    'x2': 15
}

# Define the optimal values of the variables
optimal_values = {
    'x1': x1.value,
    'x2': x2.value
}


# Calculate the total profit
total_profit = sum([profit_per_unit[var] * optimal_values[var] for var in optimal_values])

print("Total Profit:", total_profit)


Status: optimal
x1 = 17.00
x2 = 5.00
Total Profit: 245.0


In [18]:
# CIB Site:y1 (standard), y2 (deluxe)
# Decision variables
y1 = cp.Variable(integer=True)
y2 = cp.Variable(integer=True)

# Define the objective function
objective = cp.Maximize(10* y1 + 15*y2)

# Define the constraints
constraints = [
    5*y1 + 3*y2 <= 60,
    5*y1 + 6*y2 <= 75,
    4*(y1 + y2) <= 120,
    y1 >= 0,
    y2 >= 0
]

# Define the optimization problem
problem = cp.Problem(objective, constraints)

# Solve the problem
problem.solve()

# Print the status of the solution
print("Status:", problem.status)

# Print the optimal values of the variables
print("y1 = {:.2f}".format(y1.value))
print("y2 = {:.2f}".format(y2.value))

# Define the profit per unit for each product
profit_per_unit = {
    'y1': 10,
    'y2': 15
}


# Define the optimal values of the variables
optimal_values = {
    'y1': y1.value,
    'y2': y2.value
}

# Calculate the total profit
total_profit = sum([profit_per_unit[var] * optimal_values[var] for var in optimal_values])

print("Total Profit:", total_profit)


Status: optimal
y1 = 3.00
y2 = 10.00
Total Profit: 180.0


3. Now, write the optimization model for the joint optimization of both sites, and provide the optimal solution and profit levels.

In [19]:
import cvxpy as cp

# BIC Site: x1 (standard), x2 (deluxe)
# CIB Site:y1 (standard), y2 (deluxe)
# Decision variables
x1 = cp.Variable(integer=True)
x2 = cp.Variable(integer=True)
y1 = cp.Variable(integer=True)
y2 = cp.Variable(integer=True)

# Define the objective function
objective = cp.Maximize(10*(x1 + y1) + 15*(x2 + y2))

# Define the constraints
constraints = [
    4*x1 + 2*x2 <= 80,
    2*x1 + 5*x2 <= 60,
    4*(x1 + x2) <= 75,
    5*y1 + 3*y2 <= 60,
    5*y1 + 6*y2 <= 75,
    4*(y1 + y2) <= 45,
    4*(x1 + x2 + y1 + y2) <= 120,
    x1 >= 0,
    x2 >= 0,
    y1 >= 0,
    y2 >= 0
]

# Define the optimization problem
problem = cp.Problem(objective, constraints)

# Solve the problem
problem.solve()

# Print the status of the solution
print("Status:", problem.status)

# Print the optimal values of the variables
print("x1 = {:.2f}".format(x1.value))
print("x2 = {:.2f}".format(x2.value))
print("y1 = {:.2f}".format(y1.value))
print("y2 = {:.2f}".format(y2.value))

# Define the profit per unit for each product
profit_per_unit = {
    'x1': 10,
    'x2': 15,
    'y1': 10,
    'y2': 15
}

# Define the optimal values of the variables
optimal_values = {
    'x1': x1.value,
    'x2': x2.value,
    'y1': y1.value,
    'y2': y2.value
}

# Calculate the total profit
total_profit = sum([profit_per_unit[var] * optimal_values[var] for var in optimal_values])

print("Total Profit:", total_profit)



Status: optimal
x1 = 10.00
x2 = 8.00
y1 = 0.00
y2 = 11.00
Total Profit: 385.0


4. Please comment on the results? How do they compare to the results obtained in Question 2.

There is some relocation on standard and delux after combining both BIC and CIB sites. For instance, standard for CIB site is optimized to 0 for the later situation.
Partial the reason is that if each site operate alone, it can use all the 120 kg raw material. However, when two sites opereate togther, BIC gets 75 and CIB only gets 45 kg.

5. In one paragraph (no more than 300 words), detail the recommendations you would provide to the CEO of the company to improve the firm’s operations?

I recommend the utilizing both sites because this scenario gives you the most profit which is 385. 
BIC Site alone generates 245 and CIB Site along generats 180. 

Exercise 2:
1. According to the regression equation given above, which variable positively affect Profitability? Which variable negatively affect Profitability? Does this intuitively make sense? Why?

 39.05 − (5.41 × 𝑆𝑡𝑎𝑡𝑒 𝑃𝑜𝑝𝑢𝑙𝑎𝑡𝑖𝑜𝑛 𝑝𝑒𝑟 𝐼𝑛𝑛) + (5.86 × 𝑃𝑟𝑖𝑐𝑒 𝑜𝑓 𝑡h𝑒 𝐼𝑛𝑛)
− (3.09 × 𝑆𝑞𝑢𝑎𝑟𝑒 𝑅𝑜𝑜𝑡 𝑜𝑓 𝑡h𝑒 𝑀𝑒𝑑𝑖𝑎𝑛 𝐼𝑛𝑐𝑜𝑚𝑒 𝑖𝑛 𝑡h𝑒 𝑎𝑟𝑒𝑎) + (1.75 × 𝐶𝑜𝑙𝑙𝑒𝑔𝑒 𝑆𝑡𝑢𝑑𝑒𝑛𝑡𝑠 𝑖𝑛 𝑡h𝑒 𝐴𝑟𝑒𝑎)




𝐶𝑜𝑙𝑙𝑒𝑔𝑒 𝑆𝑡𝑢𝑑𝑒𝑛𝑡𝑠 𝑖𝑛 𝑡h𝑒 𝐴𝑟𝑒𝑎 and 𝑃𝑟𝑖𝑐𝑒 𝑜𝑓 𝑡h𝑒 𝐼𝑛𝑛 will positively affect Profitability, because their target customers are college students and higher price leads to higher profit for this hotel.
𝑆𝑡𝑎𝑡𝑒 𝑃𝑜𝑝𝑢𝑙𝑎𝑡𝑖𝑜𝑛 𝑝𝑒𝑟 𝐼𝑛𝑛,  𝑆𝑞𝑢𝑎𝑟𝑒 𝑅𝑜𝑜𝑡 𝑜𝑓 𝑡h𝑒 𝑀𝑒𝑑𝑖𝑎𝑛 𝐼𝑛𝑐𝑜𝑚𝑒 𝑖𝑛 𝑡h𝑒 𝑎𝑟𝑒𝑎 negatively affect Profitability, probably because 𝑆𝑡𝑎𝑡𝑒 𝑃𝑜𝑝𝑢𝑙𝑎𝑡𝑖𝑜𝑛 𝑝𝑒𝑟 𝐼𝑛𝑛 increases, people will tend to stay home more and 𝑆𝑞𝑢𝑎𝑟𝑒 𝑅𝑜𝑜𝑡 𝑜𝑓 𝑡h𝑒 𝑀𝑒𝑑𝑖𝑎𝑛 𝐼𝑛𝑐𝑜𝑚𝑒 𝑖𝑛 𝑡h𝑒 𝑎𝑟𝑒𝑎 increases will decrease the target customer(could belower budget) and they will choose more expensive hotel. 

In [30]:
import pandas as pd
df = pd.read_excel('file:///Users/yujin/Downloads/Lescargot%20(1).xlsx', header=2)

df.head()

/Users/yujin/anaconda3/lib/python3.11/site-packages/openpyxl/worksheet/_read_only.py:79: UserWarning: Unknown extension is not supported and will be removed
  for idx, row in parser.parse():


Hotel            Location     Price  Price (normalized)  \
0      1  Eureka, California   2925000           -0.301823   
1      2  Fresno, California  10000000            1.699076   
2      3  Fresno, California   3750000           -0.068503   
3      4  Fresno, California   3500000           -0.139206   
4      5  Fresno, California    325000           -1.037136   

   Square Root of Median Income (normalized)  \
0                                  -0.812780   
1                                  -0.408199   
2                                  -0.408199   
3                                  -0.408199   
4                                  -0.408199   

   College Students in Area (normalized)  \
0                              -0.536413   
1                               0.311669   
2                               0.311669   
3                               0.311669   
4                               0.311669   

   State Population Per Inn (normalized)  
0                              -0.995987  
1                              -0.474279  
2                              -0.474279  
3                              -0.474279  
4                              -0.474279

2. Using this regression equation, L’Escargot created a spreadsheet model to predict profitability. L’Escargot collected data for several locations in California, which is provided in the excel spreadsheet on Canvas “L’Escargot”. Using this spreadsheet, compute the profitability for each hotel. Which one has the highest profitability? Which one has the lowest profitability?

Hotel 2 has the highest profitability with 53.38M and hotel 8 has the lowest profitability with 23.45M.



In [35]:
# Calculate profitability using the provided formula
df['Profitability'] = (
    39.05 -
    (5.41 * df['State Population Per Inn (normalized)']) +
    (5.86 * df['Price (normalized)']) -
    (3.09 * df['Square Root of Median Income (normalized)']) +
    (1.75 * df['College Students in Area (normalized)'])
)

# Display profitability for each hotel
df[['Hotel', 'Location', 'Profitability']]

# Find the hotel with the highest profitability
highest_profitability = df.loc[df['Profitability'].idxmax()]
print("Highest Profitability:")
print(highest_profitability)

# Find the hotel with the lowest profitability
lowest_profitability = df.loc[df['Profitability'].idxmin()]
print("\nLowest Profitability:")
print(lowest_profitability)

Highest Profitability:
Hotel                                                         2
Location                                     Fresno, California
Price                                                  10000000
Price (normalized)                                     1.699076
Square Root of Median Income (normalized)             -0.408199
College Students in Area (normalized)                  0.311669
State Population Per Inn (normalized)                 -0.474279
Profitability                                         53.379192
Name: 1, dtype: object

Lowest Profitability:
Hotel                                                              8
Location                                     Los Angeles, California
Price                                                        1750000
Price (normalized)                                         -0.634128
Square Root of Median Income (normalized)                   0.167686
College Students in Area (normalized)                       3.106215
State

L’Escargot has a budget of $10 million to spend on hotels. Suppose we use a “greedy” approach where we select the most profitable hotels until we ran out of budget. So we would start by the most profitable, and then if we had enough budget left, we would buy the hotel we predict to be the second most profitable, and so on.
a. Describe what we would do with this approach, i.e., which hotels would we purchase? 
All the money will go to hotel 2 since the price is 10M and we will use up all the budget.

b. What would our total predicted profitability be? (This is the sum of the predicted
profitability of all hotels we purchase.
The profit would be 53.38M which all comes from hotel 2.

c. If we are trying to maximize our total predicted profitability, is this a good approach? How
about if we were trying to maximize the average predicted profitability of the hotels we
select? How about if we had a budget of $20 million instead of $10 million?
“greedy” approach is not the best option for  maximize our total predicted profitability for both 10M and 20M budget, because when it comes to the 20M budegt, we can purchase hotel 2, 6, 16 and the profit went up from 53.38M to 139.86 which is more than doubled. we need to consider not only the profit but also the the cost of purchasing hotel. 

In [41]:
def purchase_hotels(df, budget):
    # Sort hotels by profitability in descending order
    sorted_hotels = df.sort_values(by='Profitability', ascending=False)
    
    purchased_hotels = []
    total_spent = 0
    total_profitability = 0

    for _, hotel in sorted_hotels.iterrows():
        if total_spent + hotel['Price'] <= budget:
            purchased_hotels.append(hotel)
            total_spent += hotel['Price']
            total_profitability += hotel['Profitability']

    # Calculate average profitability of purchased hotels
    if purchased_hotels:
        average_profitability = total_profitability / len(purchased_hotels)
    else:
        average_profitability = 0

    return purchased_hotels, total_profitability, average_profitability, total_spent

# Try with $10 million budget
hotels_10m, total_profit_10m, avg_profit_10m, total_spent_10m = purchase_hotels(df, 10000000)

# Try with $20 million budget
hotels_20m, total_profit_20m, avg_profit_20m, total_spent_20m = purchase_hotels(df, 20_000_000)

# Output the results
print("With $10 million budget:")
print("Total Predicted Profitability:", total_profit_10m)
print("Average Predicted Profitability:", avg_profit_10m)
print("Purchased Hotels:", [hotel['Hotel'] for hotel in hotels_10m])
print("Total Spent:", total_spent_10m)

print("\nWith $20 million budget:")
print("Total Predicted Profitability:", total_profit_20m)
print("Average Predicted Profitability:", avg_profit_20m)
print("Purchased Hotels:", [hotel['Hotel'] for hotel in hotels_20m])
print("Total Spent:", total_spent_20m)

With $10 million budget:
Total Predicted Profitability: 53.379192308346
Average Predicted Profitability: 53.379192308346
Purchased Hotels: [2]
Total Spent: 10000000

With $20 million budget:
Total Predicted Profitability: 139.863380962872
Average Predicted Profitability: 46.621126987623995
Purchased Hotels: [2, 6, 16]
Total Spent: 19700000


4. Now, build an optimization model to select hotels given the $10 million budget.
a. Write out the optimization problem. Make sure to detail the decision variables, the objective function and the constraints.
Constraints is the budget less than 10M and objective function is to max the profit.

b. What is the optimal solution? Give the values of the decision variables and the optimal objective function value.
The optimal solution is to purchase hotel 5, 10, 11, 12, 13, 15, 16 with 269.92M profit and 9,800,000 spent on budget.

c. Does the optimal solution make sense intuitively? How does it compared to the greedy solution?
optimal solution make sense intuitively and is better than greedy solution because we need to consider the hotel price in addtion to the profit itself.

In [45]:
import cvxpy as cp
import numpy as np

# Calculate Profitability using the regression formula provided
df['Profitability'] = (
    39.05 -
    (5.41 * df['State Population Per Inn (normalized)']) +
    (5.86 * df['Price (normalized)']) -
    (3.09 * df['Square Root of Median Income (normalized)']) +
    (1.75 * df['College Students in Area (normalized)'])
)

# Convert DataFrame columns to numpy arrays for cvxpy
costs = df['Price'].values
profitabilities = df['Profitability'].values
hotels = df['Hotel'].values
budget = 10_000_000

# Define the optimization variables
x = cp.Variable(len(costs), boolean=True)

# Define the objective function
objective = cp.Maximize(cp.sum(profitabilities * x))

# Define the constraints
constraints = [cp.sum(costs * x) <= budget]

# Set up the problem
problem = cp.Problem(objective, constraints)

# Solve the problem
problem.solve()


# Calculate total price spent
total_price_spent = sum(costs[i] * int(x.value[i]) for i in range(len(costs)))

# Output the results
print("Status:", problem.status)
print("Total Profitability:", problem.value)
print("Total Price Spent:", total_price_spent)
selected_hotels = [hotels[i] for i in range(len(x.value)) if int(x.value[i]) == 1]
print("Selected Hotels:", selected_hotels)


Status: optimal
Total Profitability: 269.9246813771899
Total Price Spent: 9800000
Selected Hotels: [5, 10, 11, 12, 13, 15, 16]


/Users/yujin/anaconda3/lib/python3.11/site-packages/cvxpy/expressions/expression.py:621: UserWarning: 
This use of ``*`` has resulted in matrix multiplication.
Using ``*`` for matrix multiplication has been deprecated since CVXPY 1.1.
    Use ``*`` for matrix-scalar and vector-scalar multiplication.
    Use ``@`` for matrix-matrix and matrix-vector multiplication.
    Use ``multiply`` for elementwise multiplication.
This code path has been hit 7 times so far.

  warnings.warn(msg, UserWarning)
/Users/yujin/anaconda3/lib/python3.11/site-packages/cvxpy/expressions/expression.py:621: UserWarning: 
This use of ``*`` has resulted in matrix multiplication.
Using ``*`` for matrix multiplication has been deprecated since CVXPY 1.1.
    Use ``*`` for matrix-scalar and vector-scalar multiplication.
    Use ``@`` for matrix-matrix and matrix-vector multiplication.
    Use ``multiply`` for elementwise multiplication.
This code path has been hit 8 times so far.

  warnings.warn(msg, UserWarning)


5. L’Escargot thinks that buying too many hotels in one city is probably not a good idea and would prefer to diversify across as many cities as possible. Add constraint(s) to your model to limit the number of hotels purchased in any city to at most 2.
a. What are the constraints that you need to add to the model? Intuitively, do you expect the new optimal objective function value to be larger, smaller or the same as before?
The new constraints is each hotel location is less than or equal to 2. I intuitively expect the new optimal objective function value to be smaller, because we might miss out the oppotunity to choose higher profit hotel due to location constraint. 

b. Write the new optimization model.

c. Solve the new model. Give the values of the decision variables and the optimal objective
function value. How does this compare to the previous solution?
The profit went lower to 205.7M with total Price Spent of 9,725,000. Hotel 1, 5, 7, 8, 10, 11 becomes the optimal choices and they locate in Eureka, Fresno, Los Angeles, Los Angeles, South Lake Tahoe, and South Lake Tahoe respectively.

In [49]:
# Convert DataFrame columns to numpy arrays for cvxpy
costs = df['Price'].values
profitabilities = df['Profitability'].values
hotels = df['Hotel'].values
Location=df['Location'].values
locations = df['Location'].values
unique_locations = df['Location'].unique()
budget = 10_000_000

# Define the optimization variables
x = cp.Variable(len(costs), boolean=True)

# Define the objective function
objective = cp.Maximize(cp.sum(profitabilities * x))

# Define the constraints
constraints = [cp.sum(costs * x) <= budget]

# Add diversification constraints
for location in unique_locations:
    indices = [i for i in range(len(locations)) if locations[i] == location]
    constraints.append(cp.sum(x[indices]) <= 2)

# Set up the problem
problem = cp.Problem(objective, constraints)

# Solve the problem
problem.solve()

# Calculate total price spent
total_price_spent = sum(costs[i] * int(x.value[i]) for i in range(len(costs)))

# Output the results
print("Status:", problem.status)
print("Total Profitability:", problem.value)
selected_hotels = [hotels[i] for i in range(len(x.value)) if int(x.value[i]) == 1]
selected_Locations = [Location[i] for i in range(len(x.value)) if int(x.value[i]) == 1]
print("Total Price Spent:", total_price_spent)
print("Selected Hotels:", selected_hotels)
print("Selected Locations:", selected_Locations)





Status: optimal
Total Profitability: 205.70090440379596
Total Price Spent: 9725000
Selected Hotels: [1, 5, 7, 8, 10, 11]
Selected Locations: ['Eureka, California', 'Fresno, California', 'Los Angeles, California', 'Los Angeles, California', 'South Lake Tahoe, California', 'South Lake Tahoe, California']


/Users/yujin/anaconda3/lib/python3.11/site-packages/cvxpy/expressions/expression.py:621: UserWarning: 
This use of ``*`` has resulted in matrix multiplication.
Using ``*`` for matrix multiplication has been deprecated since CVXPY 1.1.
    Use ``*`` for matrix-scalar and vector-scalar multiplication.
    Use ``@`` for matrix-matrix and matrix-vector multiplication.
    Use ``multiply`` for elementwise multiplication.
This code path has been hit 15 times so far.

  warnings.warn(msg, UserWarning)
/Users/yujin/anaconda3/lib/python3.11/site-packages/cvxpy/expressions/expression.py:621: UserWarning: 
This use of ``*`` has resulted in matrix multiplication.
Using ``*`` for matrix multiplication has been deprecated since CVXPY 1.1.
    Use ``*`` for matrix-scalar and vector-scalar multiplication.
    Use ``@`` for matrix-matrix and matrix-vector multiplication.
    Use ``multiply`` for elementwise multiplication.
This code path has been hit 16 times so far.

  warnings.warn(msg, UserWarning)


6. In one paragraph (no more than 300 words), describe how you would present your results to L’Escargot. Do you have any recommendations for them to improve the regression model? How
about to improve the optimization model?
"Greedy" strategy and fucusing on one property is not a good option because it ignores the cost of acquiring the property and other lower cost options. It's better to optimize the portfolio with multiple objectives and constraints in order to max the profits. 

I would recommend add more variables to the regression such as seasons, week/weekend because those can affect price a lot and make the model more accurate. 
For improving the optimization model, I recommend add alternative or more contraint such as threshhold on ROI.

Question 3:

1. Assume all of the individuals in the data set would like to be matched to people of the opposite gender. What matches would you choose to maximize the aggregate match score?

The maximum aggregate match score is: 48.0
Match Laura with Carl
Match Jennifer with Adam
Match Emma with Daniel
Match Olivia with Peter
Match Martha with Eric
Match Isabella with Ralph
Match Mary with James
Match Sophia with Robert

In [51]:
import pandas as pd
import numpy as np
import cvxpy as cvx

# Load data
data_url = 'https://raw.githubusercontent.com/ormarketing/OD/master/OD.csv'
df = pd.read_csv(data_url)

# Separate males and females
females = df[df['Gender'] == 'Female']
males = df[df['Gender'] == 'Male']

# Initialize variables
num_females = len(females)
num_males = len(males)
activities = ['Sports', 'Theater', 'Religion', 'Outdoors', 'Eating Out']

# Create compatibility score matrix
compatibility_matrix = np.zeros((num_females, num_males))

for i in range(num_females):
    for j in range(num_males):
        # Calculate the compatibility score
        compatibility_matrix[i, j] = (females.iloc[i][activities] * males.iloc[j][activities]).sum()


# Optimization variables
X = cvx.Variable((num_females, num_males), boolean=True)

# Objective: Maximize sum of compatibility scores
objective = cvx.Maximize(cvx.sum(cvx.multiply(X, compatibility_matrix)))

# Constraints
# Each female is matched exactly once
constraints = [cvx.sum(X, axis=1) == 1]

# Each male is matched at most once
constraints += [cvx.sum(X, axis=0) <= 1]

# Solve the problem
problem = cvx.Problem(objective, constraints)
problem.solve()

# Output results
print("The maximum aggregate match score is:", problem.value)
matched_pairs = np.argwhere(X.value > 0.9)
for index in matched_pairs:
    female_name = females.iloc[index[0]]['Unnamed: 0']
    male_name = males.iloc[index[1]]['Unnamed: 0']
    print(f"Match {female_name} with {male_name}")


The maximum aggregate match score is: 48.0
Match Laura with Carl
Match Jennifer with Adam
Match Emma with Daniel
Match Olivia with Peter
Match Martha with Eric
Match Isabella with Ralph
Match Mary with James
Match Sophia with Robert


In [52]:
df

Unnamed: 0  Gender  Sports  Theater  Religion  Outdoors  Eating Out
0       Laura  Female       0        2        -1         0           2
1    Jennifer  Female       0        2        -1        -1           1
2        Emma  Female       2       -1         2         0           2
3      Olivia  Female       0        2         0         1           1
4      Martha  Female       0       -1        -1         0           0
5    Isabella  Female      -2        1         1         1          -2
6        Mary  Female      -2        0         2        -1           2
7      Sophia  Female       2        0        -1        -2          -2
8       James    Male       0       -2         2        -1           0
9      Robert    Male       2        1         1        -1           0
10      Peter    Male      -1        2         1         0           0
11       Eric    Male       1       -2        -2        -2           2
12     Daniel    Male       2        0         1         1           1
13       Adam    Male       2        0        -2        -2           2
14       Carl    Male       1        2        -2         2           1
15      Ralph    Male       0        2         2         2          -1

2. How would the results change if each person is presented with two possible matches?
The score is lower. I need to divide the total score by 2 since we doubled the matches which result in 41.

The maximum aggregate match score is: 82.0
Match Laura with Adam
Match Laura with Carl
Match Jennifer with Robert
Match Jennifer with Adam
Match Emma with James
Match Emma with Daniel
Match Olivia with Carl
Match Olivia with Ralph
Match Martha with Eric
Match Martha with Daniel
Match Isabella with Peter
Match Isabella with Ralph
Match Mary with James
Match Mary with Peter
Match Sophia with Robert
Match Sophia with Eric

In [54]:
df = pd.read_csv(data_url)

# Separate males and females
females = df[df['Gender'] == 'Female']
males = df[df['Gender'] == 'Male']

# Initialize variables
num_females = len(females)
num_males = len(males)
activities = ['Sports', 'Theater', 'Religion', 'Outdoors', 'Eating Out']

# Create compatibility score matrix
compatibility_matrix = np.zeros((num_females, num_males))
for i in range(num_females):
    for j in range(num_males):
        # Calculate the compatibility score
        compatibility_matrix[i, j] = (females.iloc[i][activities] * males.iloc[j][activities]).sum()

# Optimization variables
X = cvx.Variable((num_females, num_males), boolean=True)

# Objective: Maximize sum of compatibility scores
objective = cvx.Maximize(cvx.sum(cvx.multiply(X, compatibility_matrix)))

# Constraints
# Each male is matched exactly twice
constraints = [cvx.sum(X, axis=1) == 2] 

# Each female is matched at most twice
constraints += [cvx.sum(X, axis=0) == 2]

# Solve the problem
problem = cvx.Problem(objective, constraints)
problem.solve()

# Output results
print("The maximum aggregate match score is:", problem.value)
matched_pairs = np.argwhere(X.value > 0.9)
for index in matched_pairs:
    female_name = females.iloc[index[0]]['Unnamed: 0']
    male_name = males.iloc[index[1]]['Unnamed: 0']
    print(f"Match {female_name} with {male_name}")


The maximum aggregate match score is: 82.0
Match Laura with Adam
Match Laura with Carl
Match Jennifer with Robert
Match Jennifer with Adam
Match Emma with James
Match Emma with Daniel
Match Olivia with Carl
Match Olivia with Ralph
Match Martha with Eric
Match Martha with Daniel
Match Isabella with Peter
Match Isabella with Ralph
Match Mary with James
Match Mary with Peter
Match Sophia with Robert
Match Sophia with Eric


all the codes above are generated with help from chatgtp and llama3 and class code from jas 

prompt: all the questions from the hw and additional follow ups:

can we use the import cvxpy as cp package?
please match the hotel as well in the answer
i also want to know the total price 


Highest Profitability:
Hotel                                                         2
Location                                     Fresno, California
Price                                                  10000000
Price (normalized)                                     1.699076
Square Root of Median Income (normalized)             -0.408199
College Students in Area (normalized)                  0.311669
State Population Per Inn (normalized)                 -0.474279
Profitability                                         53.379192
Name: 1, dtype: object

Lowest Profitability:
Hotel                                                              8
Location                                     Los Angeles, California
Price                                                        1750000
Price (normalized)                                         -0.634128
Square Root of Median Income (normalized)                   0.167686
College Students in Area (normalized)                       3.106215
State